src/voxelalgorithms.cpp

```diff
@@ -153,5 +156,595 @@ SunlightPropagateResult propagateSunlight(VoxelManipulator &v, VoxelArea a,
+						}
+					}
+				}
+			} else {
+				// If this is an opaque node, it still can emit light.
+				new_light = ndef->get(n).light_source;
+			}
+
+			if (new_light > 0) {
+				light_sources.push(new_light, rel_pos, block_pos, block, 6);
+			}
+
+			if (new_light < old_light) {
+				// The node became opaque or doesn't provide as much
+				// light as the previous one, so it must be unlighted.
+
+				// Add to unlight queue
+				n.setLight(bank, 0, ndef);
+				block->setNodeNoCheck(rel_pos, n);
+				disappearing_lights.push(old_light, rel_pos, block_pos, block,
+					6);
+
+				// Remove sunlight, if there was any
+				if (bank == LIGHTBANK_DAY && old_light == LIGHT_SUN) {
+					for (s16 y = p.Y - 1;; y--) {
+						v3s16 n2pos(p.X, y, p.Z);
+
+						MapNode n2;
+
+						n2 = map->getNodeNoEx(n2pos, &is_valid_position);
+						if (!is_valid_position)
+							break;
+
+						// If this node doesn't have sunlight, the nodes below
+						// it don't have too.
+						if (n2.getLight(LIGHTBANK_DAY, ndef) != LIGHT_SUN) {
+							break;
+						}
+						// Remove sunlight and add to unlight queue.
+						n2.setLight(LIGHTBANK_DAY, 0, ndef);
+						map->setNode(n2pos, n2);
+						relative_v3 rel_pos2;
+						mapblock_v3 block_pos2;
+						getNodeBlockPosWithOffset(n2pos, block_pos2, rel_pos2);
+						MapBlock *block2 = map->getBlockNoCreateNoEx(
+							block_pos2);
+						disappearing_lights.push(LIGHT_SUN, rel_pos2,
+							block_pos2, block2,
+							4 /* The node above caused the change */);
+					}
+				}
+			} else if (new_light > old_light) {
+				// It is sure that the node provides more light than the previous
+				// one, unlighting is not necessary.
+				// Propagate sunlight
+				if (bank == LIGHTBANK_DAY && new_light == LIGHT_SUN) {
+					for (s16 y = p.Y - 1;; y--) {
+						v3s16 n2pos(p.X, y, p.Z);
+
+						MapNode n2;
+
+						n2 = map->getNodeNoEx(n2pos, &is_valid_position);
+						if (!is_valid_position)
+							break;
+
+						// This should not happen, but if the node has sunlight
+						// then the iteration should stop.
+						if (n2.getLight(LIGHTBANK_DAY, ndef) == LIGHT_SUN) {
+							break;
+						}
+						// If the node terminates sunlight, stop.
+						if (!ndef->get(n2).sunlight_propagates) {
+							break;
+						}
+						relative_v3 rel_pos2;
+						mapblock_v3 block_pos2;
+						getNodeBlockPosWithOffset(n2pos, block_pos2, rel_pos2);
+						MapBlock *block2 = map->getBlockNoCreateNoEx(
+							block_pos2);
+						// Mark node for lighting.
+						light_sources.push(LIGHT_SUN, rel_pos2, block_pos2,
+							block2, 4);
+					}
+				}
+			}
+
+		}
+		// Remove lights
+		unspreadLight(map, ndef, bank, disappearing_lights, light_sources,
+			modified_blocks);
+		// Initialize light values for light spreading.
+		for (u8 i = 0; i <= LIGHT_SUN; i++) {
+			const std::vector<ChangingLight> &lights = light_sources.lights[i];
+			for (std::vector<ChangingLight>::const_iterator it = lights.begin();
+					it < lights.end(); it++) {
```

Comment:
I've asked that in C++ once, and they only said "maybe its optimized, maybe not".

Either way, we have commit  34b7a147dcf9831f3b4d81599c473ba01ff5da00 and I'd prefer if it++ wouldn't start entering the codebase again.